pkg/controller/job/job_controller.go

```diff
@@ -1667,6 +1709,16 @@ func ensureJobConditionStatus(list []batch.JobCondition, cType batch.JobConditio
 	return list, false
 }
 
+func isPodFailed(p *v1.Pod, wFinalizers bool) bool {
+	if p.Status.Phase == v1.PodFailed {
+		return true
+	}
+	// When tracking with finalizers: counting deleted Pods as failures to
```

Comment:
Here is something else to think for beta:

When a pod has a deletion timestamp, it should already have a PodCondition for the cases of DisruptionTarget. However, it might not have exitCodes yet.

Maybe we should only check pods that are actually marked as `Phase == Failed`. But then this means that we should not declare a Job as failed until all the pods associated to it are either Failed or Succeeded.
One special case here are Pods that where never scheduled (they don't have `.spec.nodeName`). They always stay in `Phase=Pending`. We should count them as failures right away. I don't expect them to match PodConditions and/or exitCodes. And the only reason they would be deleted is if a user deleted them explicitly. Nothing in kubernetes deletes these pods automatically.

We can gate this new behavior on the PodDisruptionConditions feature gate.

This same question applies for the code in `trackJobStatusAndRemoveFinalizers`

Reply:
> Maybe we should only check pods that are actually marked as `Phase == Failed`. But then this means that we should not declare a Job as failed until all the pods associated to it are either Failed or Succeeded. 

I see no problem with declaring the job as failed as soon as there is at least one pod which is failed and matched the JobFail rule.

> One special case here are Pods that where never scheduled (they don't have `.spec.nodeName`). They always stay in `Phase=Pending`. We should count them as failures right away. I don't expect them to match PodConditions and/or exitCodes. And the only reason they would be deleted is if a user deleted them explicitly. Nothing in kubernetes deletes these pods automatically.

This was discussed as part of the KEP for Alpha what to do with pods which are not scheduled due to a configuration issue. It is stated as a non-goal for this KEP: https://github.com/kubernetes/enhancements/blob/abe9a2b99278b3667de8df7166995376cc7a8971/keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md?plain=1#L276

I think it makes sense no to couple this to the KEP which is already fairly large in scope. Maybe we should consider marking of misconfigured pods stuck in phase Pending as a separate KEP? It seems to me that once we mark such pods (which are currently stuck in Pending) as failed, along with pods deleted by PodGC, we could simplify the logic in Job controller to only check pods with phase=Failed.

Reply:
> This was discussed as part of the KEP for Alpha what to do with pods which are not scheduled due to a configuration issue. It is stated as a non-goal for this KEP:

That's kind of the point. We should be excluding those pods from matching conditions. It doesn't look like we are doing that if they have a DeletionTimestamp.

> It seems to me that once we mark such pods (which are currently stuck in Pending) as failed, along with pods deleted by PodGC, we could simplify the logic in Job controller to only check pods with phase=Failed.

I agree. I see this as part of the item that we already left for beta https://git.k8s.io/enhancements/keps/sig-apps/3329-retriable-and-non-retriable-failures#beta

> Simplify the code in job controller responsible for detection of failed pods based on the fix for pods stuck in the running phase

Reply:
Discussed offline and we have a plan for beta:

A small worker in the "disruption" controller that would look for pods in phase=Pending and with a deletionTimestamp and transition them to phase=Failed. Then, we can simplify the job controller to **only** check for pods with phase=Failed.

Again, this is already part of the beta criteria discussed in the previous comment.